### PR TITLE
RAG Toolkit: Out of KB questions [GSK-3372]

### PR DESCRIPTION
**Pull request by JiaenLiu**
## Description

First prototype of out of knowledge base question

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [x] I've written tests for all new methods and classes that I created.
- [ ] I've written the docstring in Google format for all the methods and classes that I used.
- [ ] I've updated the `pdm.lock` running `pdm update-lock` (only applicable when `pyproject.toml` has been
  modified)

### CODE REVIEW

File: giskard/rag/metrics/__init__.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 from .base import Metric
 from .correctness import CorrectnessMetric, correctness_metric
+# from .ragas_metrics import RagasMetric
```
Comment (mattbit) [inline giskard/rag/metrics/__init__.py:3]: ?

File: giskard/rag/question_generators/ookb_questions.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
...xt year, as she met friends of mine from the startup world, she was surprised how different reality was. And how colorful their stories were. So she decided to compile a book of interviews with startup founders.\n\nWhen the bank had financial problems and she had to fire half her staff, she started looking for a new job. In early 2005 she interviewed for a marketing job at a Boston VC firm. It took them weeks to make up their minds, and during this time I started telling her about all the things that needed to be fixed about venture capital. They should make a larger number of smaller investments instead of a handful of giant ones, they should be funding younger, more technical founders instead of MBAs, they should let the founders remain as CEO, and so on.\n\nOne of my tricks for writing essays had always been to give talks. The prospect of having to stand up in front of a group of people and tell them something that won't waste their time is a great spur to the imagination. When the Harvard Computer Society, the undergrad computer club, asked me to give a talk, I decided I would tell them how to start a startup. Maybe they'd be able to avoid the worst of the mistakes we'd made.")
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:68]: can we shorten this? it's just an example passed on every call, the shorter the better in terms of costs for the user

File: giskard/rag/question_generators/ookb_questions.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+QUESTION_FILTER_OUTPUT_EXAMPLE = """
+
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:71]: remove extra blank lines

File: giskard/rag/question_generators/ookb_questions.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+1. add some fake details/facts which is not inclued in the whole provided context but based on the detail/facts you selected in the previous step.
+2. Isolate this added fake detail/fact into a single sentence
+3. Generate a question based on the new added isolated detail/fact, fit the question in the context and can not be answered by the information in the context.
+
+The user will provide the context and give your answer after read it thoroughly.
+You will return the isolated detail/fact and the question based exclusively on the new added isolated context.
+You must output a single JSON object with keys 'selected detail/fact', 'fake detail/fact' and 'question' , without any other wrapping text or markdown and everything is in low letter. Make sure you only return valid JSON. 
+"""
+
+OoKB_QUESTION_EXAMPLE_INPUT = """Paul Graham liked to buy a baguette every day at the local market."""
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:23]: let's stick to uppercase "OOKB"

File: giskard/rag/question_generators/ookb_questions.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+            # check the generated question
+            check_messages = self._OoKB_question_check_prompt.to_messages(
+                system_prompt_input={"agent_description": agent_description, "language": language},
+                user_input=dict(question=generated_qa["question"], context=seed_document.content, related_context=context_str)
+            )
+
+            check_result = self._llm_complete(messages=check_messages)
+
+            print("check_result: ", check_result)
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:164]: Remove print statement

File: giskard/rag/question_generators/ookb_questions.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+        # 3 attempts to generate a question that can not be answered by the context
+        for _ in range(3):
+            # setup the ookb question generation prompt
+            question_messages = self._OoKB_question_generation_prompt.to_messages(
+                system_prompt_input={"agent_description": agent_description, "language": language},
+                user_input=seed_document.content,
+            )
+
+            print("current number of runs is: ", _)
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:152]: Remove print statement

File: giskard/rag/question_generators/ookb_questions.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+Your task is to check if the given question can be answered by the context and the related context or not. 
+In case the question is asking some facts in the context, unless the facts are explicitly mentioned in the context, the question can not be answered by the context.
+
+All the generated content should be in the following language: {language}
+
+Think step by step and consider the context in its entirety. Remember: you need to have a strong and sound reason to support your evaluation.
+If the question can be answered by the context, return True along with the reason. If the question can not be answered by the context, return False along with the reason.
+You must output a single JSON object with keys 'can_be_answered' and 'correctness_reason'. NO ANY MARKDOWN STRING. Make sure you return a valid JSON object.
+"""
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:49]: In the final form, we don't really need a reason or JSON format. We can just ask to get a single  token Y/N answer.

File: giskard/rag/question_generators/ookb_questions.py [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    )
+
+    _question_type = "out of scope"
+
+    # TODO: split the following function into two parts,
+
+    # 1. generate one single question by using OoKB_PROMPT with example input and output
+    # 2. check the question by using QUESTION_FILTER_INPUT_TEMPLATE with example input and output
+    # 3. if the question can not be answered by the context, return the Question Object with the DUMMY_ANSWER
+    # 4. if the question can be answered by the context, back to step 1 and generate a new question and check it again until the question can not be answered by the context.
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:120]: ??

File: tests/rag/test_question_generators.py [hunk 9]
```diff
@@ -285,3 +285,48 @@ def test_conversational_question_generation():
     ]
     assert question["metadata"]["question_type"] == "conversational"
     assert question["metadata"]["seed_document_id"] == 2
+
+def test_ookb_question_generation():
+    knowledge_base = Mock()
+    llm_client = Mock()
+    llm_client.complete = Mock()
```
Comment (mattbit) [inline tests/rag/test_question_generators.py:292]: no need, attributes of mocks are already mocks

File: giskard/rag/question_generators/ookb_questions.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+    def generate_single_question(self, knowledge_base: KnowledgeBase, agent_description: str, language: str) -> dict:
+        """
+        Generate a question from a list of context documents.
+
+        Parameters
+        ----------
+        context_documents : Sequence[Document]
+            The context documents to generate the question from.
```
Comment (mattbit) [inline giskard/rag/question_generators/ookb_questions.py:129]: wrong

**Review by mattbit — DISMISSED** [review]

File: giskard/rag/question_generators/ookb_questions.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    "can_be_answered": false,
+    "correctness_reason": "The name of the Boston investment bank where Jessica worked is not directly mentioned in the context."
+}
+
+"""
+
+DUMMY_ANSWER = "This question can not be answered by the context. No sufficient information is provided in the context to answer this question."
+
+
+class OutofKnowledgeBaseGenerator(GenerateFromSingleQuestionMixin, _LLMBasedQuestionGenerator):
```
[the remark this answers is not in the document]
Comment (pierlj) [reply giskard/rag/question_generators/ookb_questions.py:82]: `UnanswerableQuestionsGenerator` maybe. It's not much shorter but I think it is more intelligible

File: giskard/rag/question_generators/ookb_questions.py (hunk 2 above)
Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py:68]: I think it is possible by just giving one paragraph.

File: giskard/rag/metrics/__init__.py (hunk 1 above)
Comment (JiaenLiu) [reply giskard/rag/metrics/__init__.py:3]: Some redundant code found during the test and not deleted.

File: giskard/rag/question_generators/ookb_questions.py (hunk 7 above)
Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py:49]: But I am not sure without the reason, this could cause the problem of lazyness and reduce the accuracy of LLM's response.

File: giskard/rag/question_generators/ookb_questions.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py:82]: I think this name could cause some confusion :( but I do not have any ideas about the name.

File: giskard/rag/question_generators/ookb_questions.py (hunk 8 above)
Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py]: Removed, some thoughts during the development.

File: giskard/rag/question_generators/ookb_questions.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+from .base import GenerateFromSingleQuestionMixin, _LLMBasedQuestionGenerator
+from .prompt import QAGenerationPrompt
+
+OOKB_PROMPT = """
+You are a powerful auditor and mindful judger, your role is to generate question from a given context and 
+by adding some fake or non-existing details/facts to the context to check whether the agent model you are auditing is capable of reasoning and answering questions
+which has no direct answer in the provided context.
+
+There are your tasks, you should finish them step by step:
+0. select one detail/fact from the context
```
Comment (pierlj) [inline giskard/rag/question_generators/ookb_questions.py:13]: Minor stuff: I would index the list beginning with 1 instead of 0, it's more common.

File: giskard/rag/question_generators/ookb_questions.py [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+1. add some fake details/facts which is not inclued in the whole provided context but based on the detail/facts you selected in the previous step.
+2. Isolate this added fake detail/fact into a single sentence
+3. Generate a question based on the new added isolated detail/fact, fit the question in the context and can not be answered by the information in the context.
+
+The user will provide the context and give your answer after read it thoroughly.
+You will return the isolated detail/fact and the question based exclusively on the new added isolated context.
+You must output a single JSON object with keys 'selected detail/fact', 'fake detail/fact' and 'question' , without any other wrapping text or markdown and everything is in low letter. Make sure you only return valid JSON. 
+"""
+
+OOKB_QUESTION_EXAMPLE_INPUT = """Paul Graham liked to buy a baguette every day at the local market."""
```
Comment (pierlj) [inline giskard/rag/question_generators/ookb_questions.py:23]: Use a longer paragraph, it is unlikely that the text chunks provided will only contain 1 sentence. Then, your selected fact can remain the same but it forces the model to actually extract a fact, not simple copy/paste the context

File: giskard/rag/question_generators/ookb_questions.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+You are a powerful auditor and mindful juder, your role is to check whether the input question can be answered by the context and the related context documents or not.
+
+The agent model you are auditing is the following:
+- Agent description: {agent_description}
+
+Your task is to check if the given question can be answered by the context and the related context or not. 
+In case the question is asking some facts in the context, unless the facts are explicitly mentioned in the context, the question can not be answered by the context.
+
+All the generated content should be in the following language: {language}
+
```
Comment (pierlj) [inline giskard/rag/question_generators/ookb_questions.py:43]: Unnecessary

File: giskard/rag/question_generators/ookb_questions.py [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+Think step by step and consider the context in its entirety. Remember: you need to have a strong and sound reason to support your evaluation.
+If the question can be answered by the context, return True along with the reason. If the question can not be answered by the context, return False along with the reason.
+You must output a single JSON object with keys 'can_be_answered' and 'correctness_reason'. NO ANY MARKDOWN STRING. Make sure you return a valid JSON object.
```
Comment (pierlj) [inline giskard/rag/question_generators/ookb_questions.py:46]: Why did you add this?

File: giskard/rag/question_generators/ookb_questions.py [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+        """
+        seed_document = knowledge_base.get_random_document()
+        context_documents = knowledge_base.get_neighbors(
+            seed_document, self._context_neighbors, self._context_similarity_threshold
+        )
+        context_str = "\n------\n".join(["", *[doc.content for doc in context_documents], ""])
+
+        reference_context = "\n\n".join([f"Document {doc.id}: {doc.content}" for doc in context_documents])
+
+        for _ in range(3):
```
Comment (pierlj) [inline giskard/rag/question_generators/ookb_questions.py:138]: I would remove this retry loop, instead you could use retry decorator, but I would rather improve the prompt so that we don't need the retry at all.

Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py:138]: I am not sure whether we could achieve that but some prompt engineering is worth to do. Since current approach is quiet unstable and I have two potential improvements but not implemented and tested yet. For me I prefer we could move this retry to where this generate function be called (line 87 `giskard/rag/base.py`) instead of modifying the loop to retry decorator.

File: giskard/rag/question_generators/ookb_questions.py (hunk 15 above)
Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py:46]: To check if the generated question in previous step can be answered by the context or not, a.k.a to make sure the question is really out of knowledge base. Also prevent the LLM's lazyness.

File: giskard/rag/question_generators/ookb_questions.py (hunk 16 above)
Comment (JiaenLiu) [reply giskard/rag/question_generators/ookb_questions.py:138]: From the experiment I did before, the success rate is about 50%-60%. I could fine one or two good examples for the prompt but not sure about removing the check part.

File: giskard/rag/question_generators/ookb_questions.py (hunk 15 above)
Comment (pierlj) [reply giskard/rag/question_generators/ookb_questions.py:46]: Sorry I meant the "NO ANY MARKDOWN STRING", did you observe the model to answer with markdown strings ? The parsing should be handled properly already.

File: giskard/rag/question_generators/ookb_questions.py (hunk 16 above)
Comment (pierlj) [reply giskard/rag/question_generators/ookb_questions.py:138]: If you can get to around 80% of correct questions simply by tweaking the prompt and improving the examples, it would be sufficient. The extra check call could then be removed.

**Review by pierlj — APPROVED** [review]
